### PR TITLE
Added CDF distance to Univariate Solver

### PR DESCRIPTION
**Pull request by Daniel-Packer**
This implements a univariate OT solver for the case where points are given probability weights, following the approach in the implementation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html).

Currently, I haven't hooked it up to the Lower Bound Solver. I also haven't written any tests for it. Let me know if either of those things would be desirable in this PR.

### CODE REVIEW

**Comment by Daniel-Packer** [comment]
(I also noticed that in one of the tests there was still a comment referring to the Lower Bound solver has "Histogram Transport", so I changed that).

**Comment by marcocuturi** [comment]
thanks a lot Daniel!

indeed any PR that contributes a new feature must also include a test that
tests that feature!

best

**Comment by Daniel-Packer** [comment]
Great! I added a test that the output of the a Sinkhorn solver (with very small regularization) coincides with the univariate solver.

We could also consider adding another test to the LowerBound solver too, but I also still need to add the multi-weight option to the LowerBound solver.

File: tests/solvers/linear/univariate_test.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    b = jax.random.uniform(rngs[3], (self.m,))
+
+    #  adding zero weights to test proper handling
+    a = a.at[0].set(0)
+    b = b.at[3].set(0)
+    self.a = a / jnp.sum(a)
+    self.b = b / jnp.sum(b)
+
+  @pytest.mark.fast.with_args(p=[1.0, 2.0, 1.7])
+  def test_cdf_distance(self, p):
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:30]: p: float

File: src/ott/solvers/linear/univariate.py [hunk 2]
```diff
@@ -44,29 +44,43 @@ class UnivariateSolver:
...
+      y: jnp.ndarray,
+      a: Optional[jnp.ndarray] = None,
+      b: Optional[jnp.ndarray] = None
+  ) -> float:
     """Computes the Univariate OT Distance between `x` and `y`.
 
     Args:
       x: The first distribution of shape ``[n,]`` or ``[n, 1]``.
       y: The second distribution of shape ``[m,]`` or ``[m, 1]``.
+      a: The probability weights for `x` (only for wasserstein)
```
Comment (michalk8) [inline src/ott/solvers/linear/univariate.py:82]: Would rephrase as;
```rst
a: The first marginals when ``method = 'wasserstein'``. If :obj:`None`, uniform will be used.
```
Also in the future, this should be refactored (most likely accepting again a `LinearProblem` with `PointCloud` geometry).

File: tests/solvers/linear/univariate_test.py (hunk 1 above)
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:30]: Can you also please add a test for differentiability?

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/ott/solvers/linear/univariate.py [hunk 3]
```diff
@@ -91,12 +105,38 @@ def __call__(self, x: jnp.ndarray, y: jnp.ndarray) -> float:
...
     else:
       raise NotImplementedError(f"Method `{self.method}` not implemented.")
 
     # re-scale when subsampling
     return self.cost_fn.pairwise(xx, yy) * (n / xx.shape[0])
 
+  def _cdf_distance(
+      self, x: jnp.ndarray, y: jnp.ndarray, a: jnp.ndarray, b: jnp.ndarray
+  ):
+    # Implementation based on `scipy` implementation for
```
[the remark this answers is not in the document]
Comment (Daniel-Packer) [reply src/ott/solvers/linear/univariate.py:117]: Should I fix it or throw an error if they don't sum to 1?

File: src/ott/solvers/linear/univariate.py [hunk 4]
```diff
@@ -44,29 +44,43 @@ class UnivariateSolver:
       - `'quantile'` - Take equally spaced quantiles of the distances.
       - `'equal'` - No subsampling is performed, requires distributions to have
         the same number of points.
+      - `'wasserstein'` - Compute the `p`-norm distance between the CDFs of the
+        weighted distributions.
 
     n_subsamples: The number of samples to draw for the "quantile" or
       "subsample" methods.
+    p: The exponent for the `'wasserstein'` method.
```
[the remark this answers is not in the document]
Comment (Daniel-Packer) [reply src/ott/solvers/linear/univariate.py]: Yeah! Me too. When I was first implementing it, I thought that vmapping the `cost_fn` would create more overhead, but presumably it won't, right?

Anyway, I changed it.

File: tests/solvers/linear/univariate_test.py (hunk 1 above)
Comment (Daniel-Packer) [reply tests/solvers/linear/univariate_test.py]: changed to `cost_fn: costs.CostFn`

File: src/ott/solvers/linear/univariate.py (hunk 2 above)
Comment (Daniel-Packer) [reply src/ott/solvers/linear/univariate.py:79]: Do you mean the whole solver or this particular method?

File: src/ott/solvers/linear/univariate.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Daniel-Packer) [reply src/ott/solvers/linear/univariate.py:119]: (I've implemented it now so that it automatically normalizes them to sum to 1).

**Comment by Daniel-Packer** [comment]
Okay! I think I've done the desired changes! Let me know about the decision about ensuring `jnp.sum(a)=1`.

File: tests/solvers/linear/univariate_test.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class TestUnivariate:
+
+  @pytest.fixture(autouse=True)
+  def initialize(self, rng: jax.random.PRNGKeyArray):
+    self.rng = rng
+    self.n = 17
+    self.m = 29
+    self.rng, *rngs = jax.random.split(self.rng, 5)
```
[the remark this answers is not in the document]
Comment (michalk8) [reply tests/solvers/linear/univariate_test.py:31]: Should be `jax.random.split(self.rng, 5)`

File: src/ott/solvers/linear/univariate.py [hunk 6]
```diff
@@ -91,12 +104,52 @@ def __call__(self, x: jnp.ndarray, y: jnp.ndarray) -> float:
...
+
+    quantiles = jnp.sort(jnp.concatenate([x_cdf, y_cdf]))
+    x_cdf_inv = all_values_sorted[jnp.searchsorted(x_cdf, quantiles)]
+    y_cdf_inv = all_values_sorted[jnp.searchsorted(y_cdf, quantiles)]
+    return jnp.sum(
+        jax.vmap(self.cost_fn)(y_cdf_inv[1:, None], x_cdf_inv[1:, None]) *
+        jnp.diff(quantiles)
+    )
+
+  def _validate_distribution(self, x, p):
```
Comment (michalk8) [inline src/ott/solvers/linear/univariate.py:141]: I'd remove (also the shape check, would say not really necessary) this function and use this in the code above:
```python
a = jnp.ones_line(x) if a is None else a
a /= jnp.sum(a)
b = jnp.ones_like(y) if b is None else b
b /= jnp.sum(b)

```

File: tests/solvers/linear/univariate_test.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    univariate_solver = univariate.UnivariateSolver(
+        method="wasserstein", cost_fn=cost_fn
+    )
+    distance = univariate_solver(self.x, self.y, self.a, self.b)
+
+    geom = pointcloud.PointCloud(
+        x=self.x[:, None], y=self.y[:, None], cost_fn=cost_fn, epsilon=5e-5
+    )
+    prob = linear_problem.LinearProblem(geom, a=self.a, b=self.b)
+    sinkhorn_solver = sinkhorn.Sinkhorn(max_iterations=int(1e6))
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:62]: Why this huge number of iterations for a test? Would increase eps and not specify this, unless 100% needed (otherwise, would not mark it as fast).
Plus I'd also jit the solver.

File: tests/solvers/linear/univariate_test.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    rngs = jax.random.split(rng, 4)
+    eps, tol = 1e-4, 1e-3
+
+    solver = univariate.UnivariateSolver(method="wasserstein", cost_fn=cost_fn)
+
+    grad_x, grad_y, grad_a, grad_b = jax.jit(jax.grad(solver, (0, 1, 2, 3))
+                                            )(self.x, self.y, self.a, self.b)
+
+    # Checking geometric grads:
+    v_x = jax.random.normal(rngs[0], shape=self.x.shape)
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:100]: Please open an issue to create/use function similar to [check_grads](https://github.com/google/jax/issues/2648) and add a TODO to use it here.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/ott/solvers/linear/univariate.py (hunk 6 above)
Comment (Daniel-Packer) [reply src/ott/solvers/linear/univariate.py:141]: Sounds good to me!

File: tests/solvers/linear/univariate_test.py (hunk 7 above)
Comment (Daniel-Packer) [reply tests/solvers/linear/univariate_test.py:62]: This was to ensure that the resulting distribution is sufficiently close to the unregularized coupling. We can totally decrease `epsilon`, but it will cost us lowering the tolerance in the `assert_allclose` line

File: tests/solvers/linear/univariate_test.py (hunk 8 above)
Comment (Daniel-Packer) [reply tests/solvers/linear/univariate_test.py:102]: Done! It's issue #454.

File: tests/solvers/linear/univariate_test.py (hunk 7 above)
Comment (Daniel-Packer) [reply tests/solvers/linear/univariate_test.py]: I'll mark it not as fast, I think.

**Comment by Daniel-Packer** [comment]
@michalk8 This appears to be failing on a test that this PR isn't touching. Do you know what's up with that?

**Comment by marcocuturi** [comment]
@Daniel-Packer indeed, these failed tests are not related to your PR, but instead some recent JAX update. We're trying to find a suitable fix.

File: src/ott/solvers/linear/univariate.py [hunk 9]
```diff
@@ -91,12 +104,42 @@ def __call__(self, x: jnp.ndarray, y: jnp.ndarray) -> float:
...
     # re-scale when subsampling
     return self.cost_fn.pairwise(xx, yy) * (n / xx.shape[0])
 
+  def _cdf_distance(
+      self, x: jnp.ndarray, y: jnp.ndarray, a: Optional[jnp.ndarray],
+      b: Optional[jnp.ndarray]
+  ):
+    # Implementation based on `scipy` implementation for
+    # :func:<scipy.stats.wasserstein_distance>
+    a = jnp.ones_line(x) if a is None else a
```
Comment (michalk8) [inline src/ott/solvers/linear/univariate.py:121]: `jnp.ones_like`

File: tests/solvers/linear/univariate_test.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    univariate_solver = univariate.UnivariateSolver(
+        method="wasserstein", cost_fn=cost_fn
+    )
+    distance = univariate_solver(self.x, self.y, self.a, self.b)
+
+    geom = pointcloud.PointCloud(
+        x=self.x[:, None], y=self.y[:, None], cost_fn=cost_fn, epsilon=5e-5
+    )
+    prob = linear_problem.LinearProblem(geom, a=self.a, b=self.b)
+    sinkhorn_solver = jax.jit(sinkhorn.Sinkhorn(max_iterations=int(1e6)))
```
Comment (michalk8) [inline tests/solvers/linear/univariate_test.py:62]: I think the number is too large here still, seems like 10k should be enough, no?
Could you test locally and lower this please? If yes, would do `max_iterations=10_000`.

**Review by michalk8 — CHANGES_REQUESTED** [review]
Just small fixes, then we can merge!

File: src/ott/solvers/linear/univariate.py (hunk 9 above)
Comment (Daniel-Packer) [reply src/ott/solvers/linear/univariate.py]: Ack! My bad

File: tests/solvers/linear/univariate_test.py (hunk 10 above)
Comment (Daniel-Packer) [reply tests/solvers/linear/univariate_test.py]: 10k wasn't enough for the given epsilon. I lowered `epsilon` and raised the tolerance to make the tests run through.

**Comment by Daniel-Packer** [comment]
> @Daniel-Packer indeed, these failed tests are not related to your PR, but instead some recent JAX update. We're trying to find a suitable fix.

Ah, my bad! Sorry if I sounded snarky!

**Review by michalk8 — APPROVED** [review]

**Comment by michalk8** [comment]
LGTM @Daniel-Packer , thanks a lot!
